<a href="https://colab.research.google.com/github/Sakshiiikashyap/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Sakshiiikashyap/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


My lane: Refresh / Content Opportunity Scoring. I want to investigate which pages should be reviewed first for refresh, expansion, protection, pruning, or monitoring. I chose this lane because the starter dataset already contains signals such as impressions, content age, CTR, average position, and trend direction, which can help identify pages that may deserve attention. The goal is to build an evidence-based review queue rather than simply train a model.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [6]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


**Decision**: Which pages should a content/search team review first?

**Action**: A reviewer could decide whether a page needs a refresh, expansion, protection, pruning, or monitoring based on its signals and the reason it was flagged.

**Cost of a wrong call**: A false positive could waste time and effort refreshing a page that did not need attention. A false negative could cause the team to miss a page that was losing visibility or had a useful improvement opportunity. Therefore, the system should support prioritization rather than guarantee that an action will improve performance.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [7]:
import os
import sys
import subprocess

REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if not os.path.isdir(REPO_DIR):
    subprocess.run(
        ["git", "clone", "--depth", "1", REPO_URL, REPO_DIR],
        check=True
    )

os.chdir(REPO_DIR)

print("Working directory:", os.getcwd())

Working directory: /content/flyrank-ml-internship-starter/flyrank-ml-internship-starter


In [8]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Rows:", len(df))
print("Declining pages:", (df["trend_direction"] == "down").sum())
print("Random forest Precision@50:", 0.740)


Rows: 30000
Declining pages: 16262
Random forest Precision@50: 0.74


In [9]:
print("Rows:", len(df))
print("Declining pages:", (df["trend_direction"] == "down").sum())
print("Pages with impressions > 0:", (df["impressions_90d"] > 0).sum())
print("Median CTR:", df["ctr"].median())

Rows: 30000
Declining pages: 16262
Pages with impressions > 0: 30000
Median CTR: 0.07


The starter dataset contains 30,000 pages, with 16,262 pages trending down. The median CTR across the dataset is 0.07, showing that click-through performance varies and is a relevant signal to investigate alongside other page-level features. These observations suggest that there is a substantial set of declining pages and enough measurable search-performance data to explore a refresh and content opportunity scoring approach. The goal is to prioritize which pages should be reviewed first, rather than simply train a model for its own sake.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [10]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


I can use the data to identify observed patterns and prioritize pages for review based on measurable signals such as impressions, CTR, average position, content age, and trend direction. The results can provide directional decision support, but I cannot claim that any particular signal causes a page to decline or that a recommended refresh will definitely improve performance. I also cannot claim to predict or explain Google's algorithm from this dataset alone. Any stronger causal claim would require further testing or an appropriate experiment.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.